source/common/http/conn_pool_grid.cc

```diff
@@ -334,5 +333,50 @@ void ConnectivityGrid::onDrainReceived() {
   }
 }
 
+bool ConnectivityGrid::shouldAttemptHttp3() {
+  if (http3_status_tracker_.isHttp3Broken()) {
+    ENVOY_LOG(trace, "HTTP/3 is broken to host '{}', skipping.", host_->hostname());
+    return false;
+  }
+  if (!alternate_protocols_.has_value()) {
+    // TODO(RyanTheOptimist): Make the alternate protocols cache required.
+    ENVOY_LOG(trace, "No alternate protocols cache. Attempting HTTP/3 to host '{}'.",
+              host_->hostname());
+    return true;
+  }
+  if (host_->address()->type() != Network::Address::Type::Ip) {
+    ASSERT(false);
+    return false;
+  }
+  uint32_t port = host_->address()->ip()->port();
+  // TODO(RyanTheOptimist): Figure out how scheme gets plumbed in here.
+  AlternateProtocolsCache::Origin origin("https", host_->hostname(), port);
+  OptRef<const std::vector<AlternateProtocolsCache::AlternateProtocol>> protocols =
+      alternate_protocols_->findAlternatives(origin);
+  if (!protocols.has_value()) {
+    ENVOY_LOG(trace, "HTTP/3 is not available to host '{}', skipping.", host_->hostname());
+    return false;
+  }
+
+  for (const AlternateProtocolsCache::AlternateProtocol& protocol : protocols.ref()) {
+    // TODO(RyanTheOptimist): Handle alternate protocols which change hostname or port.
+    if (!protocol.hostname_.empty() || protocol.port_ != port) {
+      continue;
+    }
+
+    // TODO(RyanTheOptimist): Cache this mapping, but handle the supported versions list
+    // changing dynamically.
+    for (const quic::ParsedQuicVersion& version : quic::CurrentSupportedVersions()) {
+      if (quic::AlpnForVersion(version) == protocol.alpn_) {
+        ENVOY_LOG(trace, "HTTP/3 advertised for host '{}'", host_->hostname());
```

Comment:
I see that we correctly check whether we have an h3 version in common before marking h3 as enabled. But do we use this list when actually establishing the connection?

Reply:
Nope, but that's on my queue to address. I suspect it's going to be an "exciting" prospect because the newStream() method doesn't have an obvious way to pass metadata down into the connection pool. I'll definitely get it sorted out, but not yet.

Reply:
SG. Add a TODO here?